text/0000-pbft-regular-view-change.md

```diff
@@ -0,0 +1,185 @@
+- Feature Name: pbft_regular_view_changes
+- Start Date: 2018-10-11
+- RFC PR: (leave this empty)
+- Sawtooth Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC describes an extension to the PBFT implementation to mitigate the
+effect of compromised leader nodes. The RFC takes into account leaders unfairly
+ordering batches within blocks (the "unfair ordering" problem) and leaders
+intentionally not producing new blocks to stall the network (the "silent
+leader" problem).
```

Comment:
The second problem is addressed in the original Castro paper. Is this proposal intended to implement that solution or provide an different solution (see section 4.4 of Castro)?

Reply:
Section 4.4 of the paper ("View Changes") does not actually solve this problem. "View
changes are triggered by timeouts that prevent backups from waiting indefinitely for requests to execute." This timeout only starts once a request is received (in our case a block) and makes sure that it will execute at some point. It doesn't handle the case where no request (block) ever gets received.

Reply:
Ok. So long as we are addressing the client's request (i.e. transactions by a transactor) I think that is the spirit of requests in the paper. The way I read the paper, the lead validator is aggregating requests into blocks and then transmitting that aggregation of requests (the block) in the subsequent stages (preprepare, prepare, commit). The reply stage should probably then disaggregate the block to message the original clients.